File: java-client/src/main/java/com/couchbase/client/java/search/result/SearchRowLocations.java
/*
 * Copyright (c) 2016 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.couchbase.client.java.search.result;

import com.couchbase.client.core.annotation.Stability;
import com.couchbase.client.core.api.search.result.CoreSearchRowLocation;
import com.couchbase.client.core.api.search.result.CoreSearchRowLocations;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * A default implementation of a {@link SearchRowLocations}.
 *
 * @author Simon Baslé
 * @author Michael Nitschinger
 * @since 2.3.0
 */
@Stability.Volatile
public class SearchRowLocations {

    private final CoreSearchRowLocations internal;
    private int size;

    @Stability.Internal
    public SearchRowLocations(CoreSearchRowLocations locations) {
        this.internal = locations;
    }

    private static List<SearchRowLocation> convert(List<CoreSearchRowLocation> locations) {
        return locations.stream().map(SearchRowLocation::new).collect(Collectors.toList());
    }

    public List<SearchRowLocation> get(String field) {
        return convert(internal.get(field));
    }

    public List<SearchRowLocation> get(String field, String term) {
        return convert(internal.get(field, term));
    }

    public List<SearchRowLocation> getAll() {
        return convert(internal.getAll());
    }

    public List<String> fields() {
        return internal.fields();
    }

    public List<String> termsFor(String field) {
        return internal.termsFor(field);
    }

    public Set<String> terms() {
        return internal.terms();
    }

    @Override
    public String toString() {
        return internal.toString();
    }
}

File: scala-client/src/main/scala/com/couchbase/client/scala/search/result/SearchRowLocations.scala
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.couchbase.client.scala.search.result

import com.couchbase.client.core.api.search.result.{CoreSearchRowLocation, CoreSearchRowLocations}

import scala.jdk.CollectionConverters._

/**
  * Represents the locations of a search result row. Locations show
  * where a given term occurs inside of a given field.
  *
  * @since 1.0.0
  */
case class SearchRowLocations private (private val internal: CoreSearchRowLocations) {

  /** Gets all locations, for any field and term. */
  def getAll: Seq[SearchRowLocation] = {
    internal.getAll.asScala.map(l => SearchRowLocation(l))
  }

  /** Gets all locations for a given field (any term). */
  def get(field: String): Seq[SearchRowLocation] = {
    internal.get(field).asScala.map(l => SearchRowLocation(l))
  }

  /** Gets all locations for a given field and term. */
  def get(field: String, term: String): Seq[SearchRowLocation] = {
    internal.get(field, term).asScala.map(l => SearchRowLocation(l))
  }

  /** Gets all fields in these locations. */
  def fields: Seq[String] = {
    internal.fields.asScala
  }

  /** Gets all terms in these locations. */
  def terms: collection.Set[String] = {
    internal.terms.asScala
  }

  /** Gets all terms for a given field. */
  def termsFor(field: String): Seq[String] = {
    internal.termsFor(field).asScala
  }
}

/** An FTS result row location indicates at which position a given term occurs inside a given field.
  * In case the field is an array, `arrayPositions` will indicate which index/indices in the
  * array contain the term.
  *
  * @since 1.0.0
  */
case class SearchRowLocation private (private val internal: CoreSearchRowLocation) {
  def field: String = internal.field

  def term: String = internal.term

  /** The position of the term within the field, starting at 1. */
  def pos: Int = internal.pos.intValue

  /** The start offset (in bytes) of the term in the field. */
  def start: Int = internal.start.intValue

  /** The end offset (in bytes) of the term in the field. */
  def end: Int = internal.end.intValue

  /** Contains the positions of the term within any elements, if applicable. */
  def arrayPositions: Option[Seq[Int]] =
    Option(internal.arrayPositions).map(l => l.map(v => v.intValue).toSeq)
}
